Source/WebCore/Modules/mediasession/MediaMetadata.cpp

```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+    double shortEdge = size.minDimension();
+    auto aspectRatioCoefficient = shortEdge / longEdge;
+
+    if (longEdge < m_idealSize)
+        return aspectRatioCoefficient * (0.8 * (longEdge - m_minimumSize) / (m_idealSize - m_minimumSize) + 0.2);
+    return aspectRatioCoefficient * (1.0 * m_idealSize / longEdge);
+}
+
 void MediaMetadata::refreshArtworkImage()
 {
-    const auto& mediaImages = m_metadata.artwork;
-    if (mediaImages.isEmpty()) {
-        m_artworkLoader = nullptr;
+    m_artworkImageSrc = String();
+    m_artworkImage = nullptr;
+    m_artworkLoader = nullptr;
+    if (m_metadata.artwork.isEmpty())
         return;
-    }
-    if (!m_session || !m_session->document() || mediaImages[0].src == m_artworkImageSrc)
+    if (!m_session || !m_session->document())
         return;
-    // FIXME: Implement a heuristic to retrieve the "best" image.
-    m_artworkImageSrc = mediaImages[0].src;
-    m_artworkLoader = makeUnique<ArtworkImageLoader>(*m_session->document(), m_artworkImageSrc, [this](Image* image) {
-        if (!image || !image->data())
+
+    // First look into the artwork's sizes attributes to attempt to determine the best score.
+    Vector<IntSize> sizes(m_metadata.artwork.size(), [&](size_t index) -> IntSize {
+        const auto& sizes = m_metadata.artwork[index].sizes;
+        if (sizes.isEmpty())
+            return { };
+        if (equalIgnoringASCIICase(sizes, "any"_s))
+            return { m_idealSize, m_idealSize }; // We prefer image tagged with "any" size.
+        StringView stringView(sizes);
+        IntSize size;
+        for (auto element : stringView.split(' ')) {
+            auto sizeString = element.convertToASCIILowercase();
+            Vector<std::optional<uint32_t>> numbers;
+            numbers.reserveInitialCapacity(2);
+            for (auto number : sizeString.split('x'))
+                numbers.append(parseInteger<uint32_t>(number));
+            if (numbers.size() != 2 || !numbers[0] || !numbers[1])
+                return { };
+
+            IntSize newSize = { int(*numbers[0]), int(*numbers[1]) };
+            if (size.maxDimension() < newSize.maxDimension())
+                size = newSize;
+        }
+        return size;
+    });
+
+    Vector<float> scores(sizes.size(), [&](size_t index) -> float {
+        return imageDimensionsScore(sizes[index].width(), sizes[index].height());
+    });
+
+    bool iterate = true;
+    uint32_t bestIndex = 0;
+    for (size_t i = 0; i < scores.size(); i++) {
+        if (scores[i] < 0) {
+            // size provided was invalid, ignore them all and do an iterative search instead.
+            bestIndex = 0;
+            iterate = true;
+            break;
+        }
+        if (scores[i] > scores[bestIndex]) {
+            bestIndex = i;
+            iterate = false;
+        }
+    }
+
+    tryNextArtworkImage(bestIndex, iterate);
+}
+
+void MediaMetadata::tryNextArtworkImage(uint32_t index, bool iterate)
```

Comment:
Would it be simpler to have a Vector<String> of image srcs, sorted by the image score.
Then, start with the first image src (the one with the best score) and only iterate if there is no image?

Reply:
If we can sort the src, then we only retrieve one image. There's no need to keep sorted vector anymore. 

Otherwise we can't sort until we have downloaded all the images.

Reply:
yeah, it ended up being nicer that way. and it simplifies the logic as we don't need to distinguish the case where we do need to iterate for those we don't